### PR TITLE
Adds PartiQLFunction.Aggregation and updates Plugin interface

### PR DESCRIPTION
**Pull request by RCHowell**
## Relevant Issues
N/A

## Description
This PR updates the Plugin inteface to use Kotlin fields and updates Plugin to return a single factory as multiple is not needed.

This PR also introduces the experimental definition of a PartiQLFunction.Aggregation which will be used for aggregation user-defined-functions.

## Other Information
- Updated Unreleased Section in CHANGELOG: **[YES/NO]**
Yes

- Any backward-incompatible changes? **[YES/NO]**
Yes

- Any new external dependencies? **[YES/NO]**
No

- Do your changes comply with the [Contributing Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md)
  and [Code Style Guidelines](https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md)? **[YES/NO]**
Yes

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by johnedquinn — DISMISSED** [review]
LGTM. Good work!

File: partiql-cli/src/main/kotlin/org/partiql/cli/utils/ServiceLoaderUtil.kt [hunk 1]
```diff
@@ -113,12 +113,13 @@ class ServiceLoaderUtil {
             } else {
                 listOf()
             }
-            return plugins.flatMap { plugin -> plugin.getFunctions() }
+            return plugins.flatMap { plugin -> plugin.functions }
```
Comment (yliuuuu) [inline partiql-cli/src/main/kotlin/org/partiql/cli/utils/ServiceLoaderUtil.kt:116]: Nice.

File: partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt [hunk 2]
```diff
@@ -21,12 +21,16 @@ import org.partiql.spi.function.PartiQLFunctionExperimental
 /**
  * A singular unit of external logic.
  */
+@OptIn(PartiQLFunctionExperimental::class)
```
Comment (yliuuuu) [inline partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt:24]: Could this remain to be a function level annotation?

File: partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt [hunk 3]
```diff
@@ -21,12 +21,16 @@ import org.partiql.spi.function.PartiQLFunctionExperimental
...
-    public fun getConnectorFactories(): List<Connector.Factory>
 
     /**
-     * Represents custom built-in functions to be accessed during execution.
-     **/
-    @OptIn(PartiQLFunctionExperimental::class)
-    public fun getFunctions(): List<PartiQLFunction>
+     * A [Connector.Factory] is used to instantiate a connector.
+     */
+    public val factory: Connector.Factory
```
Comment (yliuuuu) [inline partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt:30]: Talked to @johnedquinn. This change looks fine to me.

File: partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt (hunk 2 above)
Comment (yliuuuu) [reply partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt:24]: Also, should we use `@PartiQLFunctionExperimental` instead to propagate the optin requirement?

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/PartiQLFunction.kt [hunk 4]
```diff
@@ -1,20 +1,74 @@
...
 /**
- * Represents a function interface that can be overridden by external teams.
- *
- * An implementation of this interface defines the behavior of the function
- * and its signature, which includes the function's names, return type, parameters,
- * determinism, and an optional description.
+ * The [PartiQLFunction] interface is used to implement user-defined-functions (UDFs).
+ * UDFs can be registered to a plugin for use in the query planner and evaluator.
  */
+@OptIn(PartiQLValueExperimental::class)
```
Comment (yliuuuu) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/PartiQLFunction.kt:11]: Can this be moved to function level?

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/Connector.kt [hunk 5]
```diff
@@ -20,10 +20,33 @@ import com.amazon.ionelement.api.StructElement
...
+        public val name: String
+
+        /**
+         * The connector factory method.
+         *
+         * @param catalogName
+         * @param config
+         * @return
+         */
+        public fun create(catalogName: String, config: StructElement? = null): Connector
```
Comment (yliuuuu) [inline partiql-spi/src/main/kotlin/org/partiql/spi/connector/Connector.kt:50]: Curious how can the config be null?

File: partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt (hunk 2 above)
Comment (RCHowell) [reply partiql-spi/src/main/kotlin/org/partiql/spi/Plugin.kt:24]: I agree it should be function level. Will update. Thank you

File: partiql-spi/src/main/kotlin/org/partiql/spi/connector/Connector.kt (hunk 5 above)
Comment (RCHowell) [reply partiql-spi/src/main/kotlin/org/partiql/spi/connector/Connector.kt:50]: See memory connector which doesn't need a config. It's just extra options to pass the factory. If you're loading from a file or something the persisting options in a struct is useful, but memory connector just closes over its additional parameters.

**Review by yliuuuu — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**